rir/src/ir/Compiler.cpp

```diff
@@ -224,112 +224,117 @@ void emitGuardForNamePrimitive(CodeStream& cs, SEXP fun) {
+void compileSimpleFor(CompilerContext& ctx, SEXP sym, SEXP seq, SEXP body,
+                      bool voidContext,
+                      std::function<void(bool compiledSeq)> compileRegularFor) {
     Match(seq) {
         Case(LANGSXP, fun, argsSexp) {
             RList args(argsSexp);
             if (fun != symbol::Colon || args.length() != 2) {
-                return false;
-            }
-
-            SEXP start = args[0];
-            SEXP end = args[1];
-            if (TYPEOF(start) != INTSXP && TYPEOF(start) != LGLSXP &&
-                (TYPEOF(start) != REALSXP || XLENGTH(start) == 0 ||
-                *REAL(start) != (int)*REAL(start))) {
-                return false;
+                compileRegularFor(false);
+                return;
             }
 
             // for(i in m:n) {
             //   ...
             // }
             // =>
-            // i' <- m
+            // m' <- m
             // n' <- n
-            // if (i' > n') {
-            //   n' <- ceil(n') - 1
-            //   diff' <- -1
-            //   gt' <- TRUE
+            // if (!colonInputEffects(m, n)) {
+            //    <regular for>
             // } else {
-            //   n' <- floor(n')
-            //   diff' <- 1
-            //   gt' <- FALSE
-            // }
-            // while ((i' > n') == gt') {
-            //   i <- i'
-            //   i' <- i' + diff'
-            //   ...
+            //   m' <- colonCastLhs(m')
+            //   n' <- colonCastRhs(m', n')
+            //   step <- colonGetStep(m', n')
+            //   i' <- m'
+            //   while (i' != n') {
+            //     i <- i'
+            //     i' <- i' + step
+            //     ...
+            //   }
             // }
 
+            SEXP start = args[0];
+            SEXP end = args[1];
             CodeStream& cs = ctx.cs();
-            BC::Label fwdBranch = cs.mkLabel();
-            BC::Label startBranch = cs.mkLabel();
 
-            // i' <- m
+            BC::Label skipRegularForBranch = cs.mkLabel();
+            BC::Label endBranch = cs.mkLabel();
+
+            // m' <- m
             compileExpr(ctx, start);
-            cs << BC::floor() << BC::ensureNamed();
+            cs << BC::force();
             // n' <- n
             compileExpr(ctx, end);
-            cs << BC::ensureNamed();
-            // if (i' > n')
-            cs << BC::dup2() << BC::gt();
-            cs.addSrc(R_NilValue);
-            cs << BC::recordTest() << BC::brfalse(fwdBranch);
-            // {
-            // n' <- ceil(n') - 1
-            cs << BC::ceil() << BC::dec() << BC::ensureNamed() << BC::swap();
-            // diff' <- -1
-            cs << BC::push(-1);
-            // gt' <- TRUE
-            cs << BC::push(R_TrueValue);
-            cs << BC::put(3) << BC::put(2) << BC::br(startBranch) << fwdBranch;
+            cs << BC::force();
+            // if (!colonInputEffects(m, n)) {
+            cs << BC::colonInputEffects();
+            cs.addSrc(seq);
```

Comment:
putting a `recordTest` here allows the compiler to speculate on dead branches more easily